python/lsst/ip/isr/linearize.py

```diff
@@ -26,226 +26,510 @@
-        """Correct for non-linearity
-
-        @param[in] image  image to be corrected (an lsst.afw.image.Image)
-        @param[in] detector  detector info about image (an lsst.afw.cameraGeom.Detector);
-                    the name, serial and number of amplifiers must match persisted data;
-                    the bbox from each amplifier is read;
-                    the linearization coefficients are ignored in favor of the persisted values
-        @param[in] log  logger (an lsst.log.Log), or None to disable logging;
-                    a warning is logged if any pixels are out of range of their lookup table
-
-        @return an lsst.pipe.base.Struct containing:
-        - numAmps number of amplifiers found
-        - numLinearized  number of amplifiers linearized (always equal to numAmps for this linearizer)
-        - numOutOfRange  number of pixels out of range of their lookup table (summed across all amps)
-
-        @throw RuntimeError if the linearity type is wrong or if the detector name, serial
-            or number of amplifiers does not match the saved data
-        """
-        self.checkDetector(detector)
-        ampInfoCat = detector.getAmplifiers()
-        numOutOfRange = 0
-        for ampInfo, rowInd, colIndOffset in zip(ampInfoCat, self._rowIndArr, self._colIndOffsetArr):
-            bbox = ampInfo.getBBox()
-            ampView = image.Factory(image, bbox)
-            tableRow = self._table[rowInd, :]
-            numOutOfRange += applyLookupTable(ampView, tableRow, colIndOffset)
+        rowInd = int(rowInd)
+        if rowInd < 0 or rowInd > numTableRows:
+            raise RuntimeError("LinearizeLookupTable rowInd=%s not in range[0, %s)" %
+                               (rowInd, numTableRows))
+
+        tableRow = table[rowInd, :]
+        numOutOfRange += applyLookupTable(image, tableRow, colIndOffset)
 
         if numOutOfRange > 0 and log is not None:
-            log.warn("%s pixels of detector \"%s\" were out of range of the linearization table",
-                     numOutOfRange, detector.getName())
-        numAmps = len(ampInfoCat)
-        return Struct(
-            numAmps=numAmps,
-            numLinearized=numAmps,
-            numOutOfRange=numOutOfRange,
-        )
+            log.warn("%s pixels were out of range of the linearization table",
+                     numOutOfRange)
+        if numOutOfRange < image.getArray().size:
+            return True, numOutOfRange
+        else:
+            return False, numOutOfRange
 
-    def checkDetector(self, detector):
-        """Check detector name and serial number, ampInfo table length and linearity type
 
-        @param[in] detector  detector info about image (an lsst.afw.cameraGeom.Detector);
+class LinearizePolynomial(LinearizeBase):
+    """Correct non-linearity with a polynomial mode.
 
-        @throw RuntimeError if anything doesn't match
+    corrImage = uncorrImage + sum_i c_i uncorrImage^(2 + i)
```

Comment:
Should we reproduce here some of the explanation that out wrote in the RFC, e.g, that is is assumed that the k0 coefficient is zero (bias subtracted) and that c1=1 as it it degenerate with variations in gain?